src/test/java/com/cloudant/tests/ViewPaginationTests.java

```diff
@@ -0,0 +1,210 @@
+ */
+
+package com.cloudant.tests;
+
+import static org.junit.Assert.assertEquals;
+
+import com.cloudant.client.api.CloudantClient;
+import com.cloudant.client.api.Database;
+import com.cloudant.client.api.views.Key;
+import com.cloudant.client.api.views.ViewResponse;
+import com.cloudant.tests.util.CheckPagination;
+import com.cloudant.tests.util.Utils;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+
+@RunWith(Parameterized.class)
+public class ViewPaginationTests {
+
+    /**
+     * Parameters for these tests so we run each test multiple times.
+     * We run with a single key or a complex key and both ascending and descending.
+     */
+    @Parameterized.Parameters(name = "Key:{0},Descending:{1}")
+    public static Iterable<Object[]> data() {
+        return Arrays.asList(new Object[][]{
+                {CheckPagination.Type.SINGLE, true},
+                {CheckPagination.Type.SINGLE, false},
+                {CheckPagination.Type.COMPLEX, false},
+                {CheckPagination.Type.COMPLEX, true},
+        });
+    }
+
+    @Parameterized.Parameter
+    public CheckPagination.Type type;
+
+    @Parameterized.Parameter(value = 1)
+    public boolean descending;
+
+    private Database db;
+    private String dbName;
+    private CloudantClient account;
+
+    @Before
+    public void setUp() {
+        account = CloudantClientHelper.getClient();
+        dbName = "view-pagination-test-" + Utils.generateUUID();
+        db = account.database(dbName, true);
+
+        db.syncDesignDocsWithDb();
+    }
+
+    @After
+    public void tearDown() {
+        account.deleteDB(dbName);
+        account.shutdown();
+    }
+
+    /**
+     * Check that we can page through a view where the number of results
+     * is an exact multiple of the number of pages. Check each page contains the documents
+     * we expect.
+     *
+     * Page forward to the last page, back to the first page, forward to the last page and back to
+     * the first page.
+     */
+    @Test
+    public void allTheWayInEachDirectionTwice() throws Exception {
+        CheckPagination.newTest(type)
+                .descending(descending)
+                .docCount(6)
+                .docsPerPage(2)
+                .pageToPages(3, 1, 3, 1)
+                .runTest(db);
+    }
+
+    /**
+     * Check that we can page through a view where the number of results
+     * is not an exact multiple of the number of pages. Check each page contains the documents
+     * we expect.
+     *
+     * Page forward to the last page, back to the first page, forward to the last page and back to
+     * the first page.
+     */
+    @Test
+    public void partialLastPageAllTheWayInEachDirectionTwice() throws Exception {
+        CheckPagination.newTest(type)
+                .descending(descending)
+                        //.useNewView(true)
```

Comment:
That one is a different instance of the same commented out code, it must have been in 2 places.